homeassistant/components/stream/recorder.py

```diff
@@ -135,14 +42,115 @@ def prepend(self, segments: list[Segment]) -> None:
+            # the lookback segments may still have some overlap with the recorder segments
+            if segment.sequence <= last_sequence:
+                return
+            last_sequence = segment.sequence
+
+            # Open segment
+            source = av.open(
+                BytesIO(segment.init + segment.get_data()),
+                "r",
+                format=SEGMENT_CONTAINER_FORMAT,
+            )
+            # Skip this segment if it doesn't have data
+            if source.duration is None:
+                source.close()
+                return
+            source_v = source.streams.video[0]
+            source_a = (
+                source.streams.audio[0] if len(source.streams.audio) > 0 else None
+            )
+
+            # Create output on first segment
+            if not output:
+                output = av.open(
+                    self.video_path + ".tmp",
+                    "w",
+                    format=RECORDER_CONTAINER_FORMAT,
+                    container_options={
+                        "video_track_timescale": str(int(1 / source_v.time_base))
+                    },
+                )
+
+            # Add output streams if necessary
+            if not output_v:
+                output_v = output.add_stream(template=source_v)
+                context = output_v.codec_context
+                context.flags |= "GLOBAL_HEADER"
+            if source_a and not output_a:
+                output_a = output.add_stream(template=source_a)
+
+            # Recalculate pts adjustments on first segment and on any discontinuity
+            # We are assuming time base is the same across all discontinuities
+            if last_stream_id != segment.stream_id:
+                last_stream_id = segment.stream_id
+                pts_adjuster["video"] = int(
+                    (running_duration - source.start_time)
+                    / (av.time_base * source_v.time_base)
+                )
+                if source_a:
+                    pts_adjuster["audio"] = int(
+                        (running_duration - source.start_time)
+                        / (av.time_base * source_a.time_base)
+                    )
+
+            # Remux video
+            for packet in source.demux():
+                if packet.dts is None:
+                    continue
+                packet.pts += pts_adjuster[packet.stream.type]
+                packet.dts += pts_adjuster[packet.stream.type]
+                packet.stream = output_v if packet.stream.type == "video" else output_a
+                output.mux(packet)
+
+            running_duration += source.duration - source.start_time
+
+            source.close()
+
+        # Write lookback segments
+        while len(self._segments) > 1:  # The last segment is in progress
+            await self._hass.async_add_executor_job(
+                write_segment, self._segments.popleft()
+            )
+        # Make sure the first segment has been added
```

Comment:
I have a couple questions about this section.
- How is this first recv different than the recv in the while loop below? I'm thinking it may have something to do with whether or not `_segments` is populated? Perhaps the pre-loop case and post-loop case can be be handled by  moving the the segments write loop into the idle loop to handle all the cases together.
- What wakes this up when the timeout happens? I'm thinking about the case where the timeout happens but there aren't any remaining parts that are being put in to wake up this method.

Reply:
This first recv makes sure that there is at least one Segment (the one that's currently having parts added to it) added to the deque. This lets us assume that the deque has one segment in the loop below, and after the recv there are 2 segments, one that's done and one that's in progress, so we write the one that's done.
Note that technically if there is slowness in processing and the writing takes longer than the segment time, the segments could actually build up, but 1) this shouldn't happen at all based on the relative speed of these actions and 2) it doesn't really matter too much if we get behind - just some extra buffer usage.
There should always be a recv after the output gets set to idle from the event issued by RecorderOutput.cleanup() after the segment is removed. On the next loop the idle will fail, so we should exit.

Reply:
OK, the recv calls and flow and loops make sense.

Is there a chance that `idle` is True, but then once `recv` starts it doesn't ever receive anything and it goes idle? I'm having trouble remembering what wakes up recv in that case.

Reply:
The `RecorderOutput` will always get removed, either by the `IdleTimer` callback or by the `worker_finished()` task from `_run_worker`.
When the `RecorderOutput` is removed, `StreamOutput.cleanup()` is called, which fires the event and sets Idle to True (the function is not async, so the order of these should not matter from an asyncio point of view, but we can switch them if desired).
So assuming you are already in the `while not self.idle` loop, you will get at least one event to trigger `recv()` and when you check `self.idle` the next time it will be `True`, so you will exit.
The other case is if you are actually before the loop. Say the worker exits in the beginning before any segment is actually pushed, The first `recv()` will actually use up the event from `cleanup()`, but `idle` will be set to `True` at the same time, so the `while not self.idle loop` is never entered.
TL;DR Basically no matter how you unroll the loop, there's always a check whether `self.idle is True` after a `recv()` call, so if you get both those things from `cleanup()`, the loop will exit (or never start).

Reply:
Thanks, the `recv` call in `cleanup` is what I was missing.